analyzer/emerald/emerald.go

```diff
@@ -0,0 +1,272 @@
+package emerald
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/iancoleman/strcase"
+	"github.com/jackc/pgx/v4"
+	oasisConfig "github.com/oasisprotocol/oasis-sdk/client-sdk/go/config"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/oasisprotocol/oasis-indexer/analyzer"
+	"github.com/oasisprotocol/oasis-indexer/analyzer/modules"
+	"github.com/oasisprotocol/oasis-indexer/analyzer/util"
+	"github.com/oasisprotocol/oasis-indexer/config"
+	"github.com/oasisprotocol/oasis-indexer/log"
+	"github.com/oasisprotocol/oasis-indexer/metrics"
+	"github.com/oasisprotocol/oasis-indexer/storage"
+	"github.com/oasisprotocol/oasis-indexer/storage/oasis"
+)
+
+const (
+	emerald = analyzer.RuntimeEmerald
+
+	emeraldMainDamaskName = "emerald_main_damask"
+)
+
+// Main is the main Analyzer for the Emerald Runtime.
+type Main struct {
+	cfg     analyzer.RuntimeConfig
+	qf      analyzer.QueryFactory
+	target  storage.TargetStorage
+	logger  *log.Logger
+	metrics metrics.DatabaseMetrics
+
+	moduleHandlers []modules.ModuleHandler
+}
+
+// NewMain returns a new main analyzer for the Emerald Runtime.
+func NewMain(cfg *config.AnalyzerConfig, target storage.TargetStorage, logger *log.Logger) (*Main, error) {
+	ctx := context.Background()
+
+	// Initialize source storage.
+	networkCfg := oasisConfig.Network{
+		ChainContext: cfg.ChainContext,
+		RPC:          cfg.RPC,
+	}
+	factory, err := oasis.NewClientFactory(ctx, &networkCfg)
+	if err != nil {
+		logger.Error("error creating client factory",
+			"err", err,
+		)
+		return nil, err
+	}
+
+	network, err := analyzer.FromChainContext(cfg.ChainContext)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := emerald.ID(network)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := factory.Runtime(id)
+	if err != nil {
+		logger.Error("error creating runtime client",
+			"err", err,
+		)
+		return nil, err
+	}
+	roundRange := analyzer.RoundRange{
+		From: uint64(cfg.From),
```

Comment:
Should the `Config`'s `From` and `To` be `uint64`?

Reply:
Ideally yes. Feel free to push back if you'd like to see this now, but the reason I didn't do it here is because then we need in our config files something that flags an analyzer as a runtime analyzer or a consensus analyzer even though that's currently intrinsic from the name of the analyzer.

It should be a safe typecast until block `2 ^ 63 - 1` so I figured it was fine to kick the can down the road for now :P

Reply:
Aight.

Reply:
The consensus layer uses `int64` everywhere, since that's what Tendermint requires. [Here](https://blog.cosmos.network/choosing-a-type-for-blockchain-height-beware-of-unsigned-integers-714804dddf1d) is their rationale for doing so.

ParaTimes, on the other hand, us `uint64` to represent height/round number. So we need something different.